csharp/src/Apache.Arrow/ArrowBuffer.BitPackedBuilder.cs

```diff
@@ -0,0 +1,268 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Apache.Arrow
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using Apache.Arrow.Memory;
+
+    public partial struct ArrowBuffer
+    {
+        /// <summary>
+        /// The <see cref="ArrowBuffer.BitPackedBuilder"/> class is a complement to <see cref="ArrowBuffer.Builder{T}"/>
+        /// and is designed for boolean fields, which are efficiently bit-packed into byte-aligned memory.
+        /// </summary>
+        public class BitPackedBuilder
+        {
+            private const int DefaultBitCapacity = 8;
+
+            /// <summary>
+            /// Gets the number of bits of current capacity.
+            /// </summary>
+            public int BitCapacity { get; private set; }
+
+            /// <summary>
+            /// Gets the number of bits currently appended.
+            /// </summary>
+            public int BitCount { get; private set; }
```

Comment:
These properties are deliberately called `BitCapacity` and `BitCount`, as opposed to `Capacity` and `Length`, to avoid any ambiguity as to what they are counting.

Reply:
I've renamed these to `Capacity` and `Length`, with the docstring making it clear that they refer to bits.